File: Convex_Hull/Sort.cpp
#include "Sort.h"
#include <cstdlib>
#include <iostream>
using namespace std;

template < typename T>
void Sort<T>::initialize()
{
    srand(time(NULL));
    for ( unsigned i = 0; i < _N; i++ )
    {
        T randint = rand()/(RAND_MAX/_N)+1; // generate random numbers from [1,_N]
        _elems.push_back(randint);
    }
}

template < typename T>
void Sort<T>::swap( T &a, T &b )
{
    T tmp = a;
    a = b;
    b = tmp;
}

template < typename T>
void Sort<T>::displayme( ostream &os ) const
{
    for ( unsigned i = 0; i < _N; i++ )
        os << _elems[i] << ", ";
    os << endl;
}

template < typename T>
bool Sort<T>::isSorted( ) const
{
    for ( unsigned i = 0; i < _N-1; i++ )
        if ( _elems[i] > _elems[i+1] )
            return false;
    return true;
}

File: Week1_3/Sort.java
import java.util.Random;
public abstract class Sort
{
    protected int N;
    protected int[] elems;

    public Sort ( int size )
    {
        N = size;
        elems = new int[N];
    }

    public void initialize( )
    {
        Random rand = new Random();

        for ( int i = 0; i < N; i++ )
        {
            int n = rand.nextInt(N) + 1;
            elems[i] = n;
        }
    }

    public void displayme( )
    {
        for ( int i = 0; i < N; i++ )
            System.out.printf("%d, ",elems[i]);
        System.out.printf("\n");
    }

    public boolean isSorted()
    {
        for ( int i = 0; i < N-1; i++ )
            if ( elems[i] > elems[i+1] )
                return false;
        return true;
    }

    public void sort() {};

    protected void swap( int i, int j )
    {
        int tmp = elems[i];
        elems[i] = elems[j];
        elems[j] = tmp;
    }
}
